src/python_testing/TC_SC_4_1.py

```diff
@@ -0,0 +1,828 @@
+                # leading zeros, and shall not exceed 3600000
+                assert_valid_sii_key(sii_key)
+
+            # *** SAI KEY ***
+            if self.supports_icd:
+                # Verify that the 'SAI' key is present and non-empty if 'supports_icd' is True
+                asserts.assert_in('SAI', txt_record.txt, "'SAI' key must be present.")
+                sai_key = txt_record.txt['SAI']
+                asserts.assert_true(sai_key, "'SAI' key is present but has no value.")
+
+                # Verify that the 'SAI' key is an unsigned integer with units of milliseconds
+                # encoded as a variable length decimal number in ASCII text, omitting any
+                # leading zeros, and shall not exceed 3600000
+                assert_valid_sai_key(sai_key)
+
+            # *** SAT KEY ***
+            # If the 'SAT' key is present:
+            if 'SAT' in txt_record.txt:
+                sat_key = txt_record.txt['SAT']
+
+                # Verify that it is non-empty
+                asserts.assert_true(sat_key, "'SAT' key is present but has no value.")
+
+                # Verify that it is an unsigned integer with units of milliseconds encoded as
+                # a variable length decimal number in ASCII text, omitting any leading zeros,
+                # and shall be less than or equal to 65535
+                assert_valid_sat_key(sat_key)
+
+                # If the 'SAT' key is present and supports_icd is true, verify
+                # that the value is equal to 'active_mode_threshold_ms'
+                if self.supports_icd:
+                    asserts.assert_equal(int(sat_key), self.active_mode_threshold_ms,
+                                         f"'SAT' key value ({sat_key}) must be equal to 'active_mode_threshold_ms' ({self.active_mode_threshold_ms})")
+
+            # Verify that the 'SAI' key is present if the 'SAT' key is present
+            asserts.assert_true(
+                'SAT' not in txt_record.txt
+                or 'SAI' in txt_record.txt,
+                "SAI key must be present if SAT key is present."
+            )
+
+            # *** D KEY ***
+            # Verify that the 'D' key is present and non-empty
+            asserts.assert_in('D', txt_record.txt, "'D' key must be present.")
+            d_key = txt_record.txt['D']
+            asserts.assert_true(d_key, "'D' key is present but has no value.")
+
+            # Verify that it is a full 12-bit discriminator encoded as a variable length
+            # decimal number in ASCII text, omitting any leading zeros, and up to 4 digits
+            assert_valid_d_key(d_key)
+
+            # *** VP KEY ***
+            # If the VP key is present
+            if 'VP' in txt_record.txt:
+                # Verify that it is non-empty
+                vp_key = txt_record.txt['VP']
+                asserts.assert_true(d_key, "'D' key is present but has no value.")
```

Comment:
```suggestion
                asserts.assert_true(vp_key, "'VP' key is present but has no value.")
```
This is inside the VP key code block, but this looks like it is looking for the D key here instead?